Class/Practise 2.2/image.cpp

```diff
@@ -0,0 +1,102 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <array>
+#include <tuple>
+#include <ranges>
+#include <print>
+
+struct Color {
+    int r, g, b;
+
+    bool operator==(const Color& other) const {
+        return r == other.r && g == other.g && b == other.b;
+    }
+};
+
+using Image = std::array<std::array<Color, 16>, 16>;
+
+bool readImage(const std::string& inputFilename, Image& image) {
+    std::ifstream file(inputFilename);
+    if (!file.is_open()) {
+        std::println("Failed to open file: {}", inputFilename);
+        return false;
+    }
+
+    std::string line;
+    for (int i = 0; i < 16; ++i) {
+        if (!std::getline(file, line)) {
+            std::println("Invalid input format: not enough lines.");
+            return false;
+        }
+
+        std::istringstream lineStream(line);
+        for (int j = 0; j < 16; ++j) {
+            char delimiter;
+            if (!(lineStream >> image[i][j].r >> delimiter >> image[i][j].g >> delimiter >> image[i][j].b) || delimiter != ',') {
+                std::println("Invalid input format at line {}, pixel {}.", i + 1, j + 1);
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void writeImage(const std::string& outputFilename, const Image& image) {
+    std::ofstream file(outputFilename);
+    for (const auto& row : image) {
+        for (const auto& pixel : row) {
+            file << pixel.r << ',' << pixel.g << ',' << pixel.b << ' ';
+        }
+        file << '\n';
+    }
+}
+
+void applyFavoriteColor(Image& image, const Color& favoriteColor) {
+    for (int i = 0; i < 16; ++i) {
+        for (int j = 0; j < 16; ++j) {
+            if (image[i][j] == favoriteColor) {
+                if (i > 0) {
+                    image[i - 1][j] = favoriteColor; // top pixel
```

Comment:
I am not sure whether the in-place edit of the image is what we really want here. In this case it will work but if requirement will be changed to fill not only left and top but also right and bottom pixels then such type of algorithm will be broken due to inplace image.

Also when you pass some object as the first parameter and using it as output e.g. modification is perform then it look that it is the same as this pointer in class implementation and therefore we should create Image class and use this function as one of its methods.

The more flexible way is to have vector of pairs with target pixel coordinates that should be filled with favourite color. Then we will generate such a list and will go through them in a separate loop to update the image. Note: here is a classical situation when we have a little bit faster code due to single loop but flexibility is lower in terms of requirements change. I would say it is always better to use flexible case until we get some concerns related to performance - if so then we should find the ways to perform an optimization.